### PR TITLE
[FLINK-11449][table] Uncouple the Expression class from RexNodes.

### PR DESCRIPTION
**Pull request by sunjincheng121**
## What is the purpose of the change

**This pull request uncouple the Expression class from RexNodes.**
As mentioned in [FLIP-32](https://cwiki.apache.org/confluence/display/FLINK/FLIP-32%3A+Restructure+flink-table+for+future+contributions).,  we have uncouple the `Expression` class from RexNodes. 
There are 6 major changes in PR, which are listed in the `Brief change log`.

## Brief change log
 -  Introduce new expression classes and visitor interface in `flink-table-common`.
 -  Introduce `PlannerExpression extends Expression` and let built-in expressions extend it. 
 -  Introduce the big visitor. 
 -  Use new expression classes in API. 
 -  Change PlannerExpression to Expression in FilterableTableSource, and move TableSource into `flink-table-common`. 
 - * Remove join(udtf) .

## Verifying this change
This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? ( not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Expression.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.apache.flink.table.expressions;
+
+import java.util.List;
+
+/**
+ * The base class for all expressions.
+ */
+public abstract class Expression {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Expression.java]: Add the `other` section is a good idea.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Literal.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	private final Object value;
+
+	private Optional<TypeInformation<?>> type;
+
+	public Literal(Object value) {
+		this.value = value;
+		this.type = Optional.empty();
+	}
+
+	public Literal(Object value, TypeInformation<?> type) {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/Literal.java]: I think unify nulls and Literals is good idea.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TableFunctionCall.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+ */
+public class TableFunctionCall extends Call {
+
+	private Optional<String[]> alias = Optional.empty();
+
+	public TableFunctionCall(TableFunctionDefinition func, List<Expression> args) {
+		super(func, args);
+	}
+
+	public TableFunctionCall alias(String[] alias) {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TableFunctionCall.java]: Yes, I think we can keep `TableFunctionCall` for now and replace it with regular `call` solution when join(table(udtf)) is removed. What do you think?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/windows.scala [hunk 4]
```diff
@@ -143,7 +143,31 @@ case class OverWindowWithOrderBy(partitionBy: Seq[Expression], orderBy: Expressi
...
+  def as(alias: Expression): OverWindow = {
+    org.apache.flink.table.api.scala.OverWindow(
+      alias, partitionBy, orderBy, UNBOUNDED_RANGE, CURRENT_RANGE)
+  }
+}
+
+/**
+  * Over window is similar to the traditional OVER SQL.
+  */
+case class OverWindow(
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/scala/windows.scala]: The Refactoring of Expression causes the parameter type of the window API under Scala to be changed from `PlannerExpression` to `APIExpression`. Since the original java window api depends on the `ExpressionParser` and the Scala API, and the ExprssionParser is not compatible with the new `APIExpression`, Need to refactor the java window api, taking `TumbleWithSize` as an example:
The original window API:
```
Class TumbleWithSize(size: Expression) {
   Def this(size: String) = 
   this(ExpressionParser.parseExpression(size))
}
```
Such an API depends on `ExpressionParser`, and the pass depends on `Calcite`.

The current window API:

Implementation of new TumbleWithSize(size) (split into two versions of Java and scala):
```
// Java
Class TumbleWithSize(size: String) {
}
// Scala
Class TumbleWithSize(size: Expression) {
}
```
Such an API no longer relies on `ExpressionParser` and no longer passes dependencies on `Calcite`.
Is that makes sense to you? or there are better suggestions ?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala [hunk 5]
```diff
@@ -414,15 +414,15 @@ abstract class BatchTableEnvironment(
     * @tparam T The type of the [[DataSet]].
     */
   protected def registerDataSetInternal[T](
-      name: String, dataSet: DataSet[T], fields: Array[Expression]): Unit = {
+      name: String, dataSet: DataSet[T], fields: Array[PlannerExpression]): Unit = {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala]: The `XXXInternal` methods are internal methods that the user does not perceive, so IMO there is no need to become an Expression? （ just reduce the change code for current PR?)

So far I have not found the benefits of PlannerExpression to inherit Expression(Except reduce the some change code), because we already have the visitor mode to establish the conversion bridge from Expression to PlannerExpression. Expression is used at the user API level, and Expression is converted to PlannerExpression at the implementation level. We can indeed inherit PlannerExpression as Expression. Can you elaborate on the benefits of doing this?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.expressions;
+
+/**
+ * The distinct aggregate expression.
+ */
+public class DistinctAgg extends UnaryExpression {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java]: Yes, we can using the `call`, but `DistinctAgg1 is created to prevent the user from continuously calling distinct, for example: `a.count.distinct.distinct`.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TableSymbol.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.expressions;
+
+/**
+ * The base interface for all symbol, such as time interval unit, time point unit
+ * are all defined based on TableSymbol.
+ */
+public interface TableSymbol {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TableSymbol.java]: Yes, your right, enum can not be inherit in Java.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/UnresolvedFieldReference.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.expressions;
+
+/**
+ * Unresolved field reference expression.
+ */
+public class UnresolvedFieldReference extends LeafExpression {
```
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/UnresolvedFieldReference.java]: How about call it `FieldReference` ?

**Comment by sunjincheng121** [comment]
Hi @twalthr Thanks for your valued comments! 

I have been fixed all comments locally in addition to the comments which I have questions, especially the Expressions has already reduced from 21 to the 9. You will see the changes after updating the PR. I would like to ask for your opinion: I am updating the existing improvements first, or  updating all the fix together after finish all the discussion?

Thanks,
Jincheng

**Comment by sunjincheng121** [comment]
Hi @twalthr  I just updated my PR according to my understanding of your comments. I splits the changes to 3 commits:

1. Move expressions package to plan.expressions and rename to PlannerExpression.
2. Using the visitor method Uncouple the PlannerExpression class from RexNodes
3. Change PlannerExpression to Expression in TableSource

Not sure if I fully understand your suggestions,  and welcome any suggestions.

Best,
Jincheng

**Comment by KurtYoung** [comment]
Thanks @sunjincheng121 for this great effort. I just quickly go through the changes, and have some architecture comments:
1. It appears to me that i can't tell the differences between `Expression` and `PlannerExpression`. What are the scenarios for both expressions? Take `FilterableTableSource` for example, currently it takes a list of `Expression` for input. But clearly the method `applyPredicate` happens during optimize phase, my feeling is we should use `PlannerExpression` during optimize. (I know `FilterableTableSource` is API and should only accept `Expression` which is also an API, but the logic seems a little weird).

2. What about we just use current `Expression` for API, by getting rid of the `toRexNode` method. The only problem is the number of expressions is quite big and keeps increasing. To solve this, we could follow the advices @twalthr had suggested, we categorize all expressions we currently have, and squash some all them. For example, we can squash "IF", "ABS" into `CallExpression`, and use `FunctionDefinition` to represent details information. Quite similar with Calcite's `RexNode` and `SqlOperator`. And for `ExpressionVisitor`, we can directly mapping `Expression` to `RexNode`, which seems to be much clearer.

**Comment by sunjincheng121** [comment]
Hi @KurtYoung thank for you comments.

In the long-term, we will only have one expression, but for now, I think we should complete the moudle refactoring of flink ASAP., and then merge Blink into Flink Master. The coexistence of Expression and PlannerExpression (the existing Expression has not changed) will greatly reduce the difficulty and workload of Blink merge. If I understand it correctly, Timo also proposed this approach in JIRA.

Regarding to drop the existing CAST, ALIAS, EQUALTO etc., more than 100 function call definitions is a big change, and when merging the Blink, we will also need to do this kind of work which introduce extra huge overhead. Therefore, I prefer to add a thin layer of Expression abstraction to the API in the current PR. After the Blink has been merged, we can unify the Expression by adding an expression visitor, e.g.: `XXX extends ExpressionVisitor[RexNode]`  which can translate an Expresison to RexNode.

In short, I expect to complete uncouple the Expression from RexNodes in the most efficient way to reduce the work of Blink merging.

What do you think ?

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java]: I think it again, `Call` is work, but for user friendly, we can using the `AggregateCall`. i.e: Add AggregateCallExpression to avoid distinct on CallExpression directly which would bring chaos on Api. This is also consistent with the previous Flink master.

**Comment by sunjincheng121** [comment]
Hi @twalthr  I updated the PR based on your comments.

More updates with the latest commit:

1. Make PlannerExpression implements Expression
2. Add 'Expression' suffix name to all Expressions, e.g., Call to CallExpression, TableReference to TableReferenceExpression.
3. Revert package path change for PlannerExpression
4. Add AggregateCallExpression to avoid distinct on CallExpression directly which would bring chaos on Api. This is also consistent with the previous flink master. 

Note: We add the AggregateCallExpression to avoid the following chaos on Api:

1. Avoid multi distinct, e.g., 'a.sum.distinct.distinct
2. Avoid wrong order, e.g., 'a.distinct.sum
3. Avoid distinct on udf, e.g., 'a.udf.distinct

Thank you again for your valuable comments and wish you a happy holiday.

Best,
Jincheng

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java]: IMHO I think we should allow `a.count.distinct.distinct` and fail during validation. Because this is also the behavior of the Java API, right? Right now, you can also do weird things like `a.count.trim.abs`. The API does not limit this as well.

[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/DistinctAgg.java]: Yes，we can not avoid the wrong pattern in Java API, we only have the chance to fail during validation.  I really change the `distinctAgg` to the `Call`, but i find that current master code, the Scala API do not allow the follows pattern:
1. multi distinct, e.g., 'a.sum.distinct.distinct
2. wrong order, e.g., 'a.distinct.sum
3. distinct on udf, e.g., 'a.udf.distinct
So, we add the `AggregateCallExpression` for Scala user. 
What do you think?

**Comment by twalthr** [comment]
@KurtYoung What you mentioned to "use FunctionDefinition to represent details information, quite similar with Calcite's RexNode/SqlOperator" is exactly what I have in mind as well as a long-term vision.

The function definition would take information of accepted parameter types or parameter type families, input and return type inference etc. However, this is future work. It would unify built-in and user-defined functions in the future and should be implemented together with the type system rework. I started with a design document for this but will postpone it for now. In the future, we won't need dedicated classes for every expression. Only a function definition and a visitor that translates common functions into RexNodes such that the optimizer can recognize them.

But for now, let's simply translate "generic expressions" such as `Call`s to actual expression classes in a big visitor pattern in order to speed up the Blink merge.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/FunctionDefinitions.java [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+	public static final FunctionDefinition IN = new FunctionDefinition("in", OTHER);
+	public static final FunctionDefinition IS_NULL = new FunctionDefinition("isNull", OTHER);
+	public static final FunctionDefinition IS_NOT_NULL = new FunctionDefinition("isNotNull", OTHER);
+	public static final FunctionDefinition IS_TRUE = new FunctionDefinition("isTrue", OTHER);
+	public static final FunctionDefinition IS_FALSE = new FunctionDefinition("isFalse", OTHER);
+	public static final FunctionDefinition IS_NOT_TRUE = new FunctionDefinition("isNotTrue", OTHER);
+	public static final FunctionDefinition IS_NOT_FALSE = new FunctionDefinition("isNotFalse", OTHER);
+	public static final FunctionDefinition IF = new FunctionDefinition("if", OTHER);
+	public static final FunctionDefinition BETWEEN = new FunctionDefinition("between", OTHER);
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/FunctionDefinitions.java]: I like the idea of categorizing the functions but we should discuss the categories. Why is "between" not a comparison? Maybe we should find a different classification such as `CORE_PREDICATE` which includes only the core functions such as: <>=!==

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/FunctionType.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+package org.apache.flink.table.expressions;
+
+/**
+ * TThe types used to identify the function.
+ */
+public enum FunctionType {
+	LOGICAL,
+	COMPARISON,
+	TABLE_FUNCTION,
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/FunctionType.java]: Why do we have `TABLE_FUNCTION` and `AGGREGATION` but not scalar function?

Comment (sunjincheng121) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/FunctionType.java]: Yes, Scalar function will be divided into finer granularity， just like you mentioned above `CORE_PREDICATE` and other types.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/FunctionDefinitions.java (hunk 9 above)
Comment (sunjincheng121) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/expressions/FunctionDefinitions.java]: Ah, I think we can enumerate according to the original classification.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TableFunctionCall.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sunjincheng121) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TableFunctionCall.java]: Yes, we will merge this PR after 1.8 branch is forked off, so I'll remove the `TableFunctionCall` and remove the `join(udtf)`.

[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/expressions/TableFunctionCall.java]: If it simplifies this PR, feel free to apply this change here. Otherwise we can also do it in a separate PR.

**Comment by sunjincheng121** [comment]
Hi Thanks for the great proposal about split the PR. @twalthr !

According to your comments, PR has been fine-grained split. Currently, each commit is self-contained and is a function-independent change. The last commit is to remove the join (udtf), most of which is to delete the code,  and fix the test case. But i also can open a new PR for last commit, if you think that `removing join(udtf)` commit needs to be submitted in a separate PR. 

Best,
Jincheng

**Comment by twalthr** [comment]
Thanks for the update @sunjincheng121. The commits look much cleaner now. In order to speed up the process I will apply the final polishing directly in a separate branch. I will come back to you with my suggestions and minor improvements shortly. We can then discuss them again. Thanks.

**Comment by sunjincheng121** [comment]
Sounds good @twalthr !

**Comment by twalthr** [comment]
FYI: I merged the first commit in 8e3985110693d2999a4b827929891ca2e6575cd2. I will continue merging the commits step by step.

**Comment by twalthr** [comment]
FYI: I merged the second commit in d1fa236bf3c8fcb49868e17d69c3578f47940baf.

**Comment by twalthr** [comment]
@flinkbot approve description
@flinkbot approve consensus

**Comment by twalthr** [comment]
FYI: I merged the third commit in a7aa58b62ccbb6e02f22ae1282e8fbc246641022.

**Comment by twalthr** [comment]
FYI: I merged table function removal commit in dcd8c74b504046802cebf278b718e4753928a260.

**Comment by sunjincheng121** [comment]
Thanks a lot for the big merging @twalthr

**Comment by sunjincheng121** [comment]
@twalthr  I find that there is also a commit here without merged [Change PlannerExpression to Expression in FilterableTableSource], How are you going to deal with this commit?

**Comment by twalthr** [comment]
@sunjincheng121 good point. Maybe we should open a follow-up issue for this. I skipped this commit in order to not break SPI implementations for now (no changes in ORC format required yet). The current translation in `master` is very inconsistent, it is using its own function catalog instance and tries to map SQL to expressions arbitrarily. I hope we can do this nicer in the future. But this is not high on the priority list. What do you think?

**Comment by sunjincheng121** [comment]
Um, OrcTableSource#applyPredicate will call `toOrcPredicate` method which using the plannerExpreasion, such as `Or`, So, the connector should dependent on `flink-table-planner` module, which we do not want. So IMHO. if we want to reduce connector user's dependences, we need this change at the current time. Just like we do not want API user to depend on `flink-table-planner` when they write job logic. is there anything I missed?

**Comment by twalthr** [comment]
I think we are on the same page about the goal. The questions is how many times do we want to break SPI implementations. But if you think we should prioritize this issue higher we can. I would like to focus on the main API so far and let SPI API as it is for now.

**Comment by sunjincheng121** [comment]
So, I'll check whether the changes are needed by port catalog.  If so I'll create a new JIRA. for this change. If it not the blocker for our FLIP-32 items, we can lower priority.
